Discovery Planning

mmccool opened the pull request:
- Some detailed notes on planning for Discovery
- Suggested structure for capturing planning for other deliverables (including TD)
- IMO MD files in github are more appropriate for capturing this content than slides

lu-zero commented on planning/Discovery/work-items.md:
```diff
@@ -0,0 +1,51 @@
+### Prioritized List
+1. Direct - [Discovery Improvements](https://w3c.github.io/wot-charter-drafts/wot-wg-2023-details.html#discovery-improvements-workitem)
+    * Needs to be broken into smaller work-items to better support prioritization.
+2. Direct - [Discovery of Thing Models](https://w3c.github.io/wot-charter-drafts/wot-wg-2023-details.html#discovery-thing-models-workitem) 
```
A ThingModel cannot be discovered since it cannot be advertised as-is. I'd rather have it to the lowest priority since it would require overhauling a bit more TM before discussing how to advertise and how to put in the directory.

relu91 replied:
I agree that this priority level for TMs discovery needs to be supported by some explanation. As far as I know, we don't have strong use cases besides the nice-to-have ability to link TMs using relative links in TDDs.

mmccool replied:
Agree this can be lower in priority.  Right now the list has not yet been prioritized, it's just in the order they appeared in the work items.   But we would need to understand the use case for putting TMs in directories.  I can think of a couple use cases but they could be satisfied in other ways.

lu-zero commented on planning/Discovery/work-items.md:
```diff
@@ -0,0 +1,51 @@
+    * Needs to be broken into smaller work-items to better support prioritization.
+2. Direct - [Discovery of Thing Models](https://w3c.github.io/wot-charter-drafts/wot-wg-2023-details.html#discovery-thing-models-workitem) 
+3. Direct - [Discovery CoAP Directory](https://w3c.github.io/wot-charter-drafts/wot-wg-2023-details.html#discovery-coap-dir-workitem)
+4. Direct - [Discovery JSON Path Query Language](https://w3c.github.io/wot-charter-drafts/wot-wg-2023-details.html#discovery-jsonpath-query-language-workitem)
```
This is more important in my opinion, it is a core functionality that makes the Directory useful in more scenarios, what is the difference with query filters though?

mmccool replied:
The problem with all the "official" query languages is that the specs are not final.  "Query filters" would be simple, fixed-function query mechanisms that would not have a dependency on a query language.  It's also a way for us to do geolocation filters without depending on extending the official query languages (there is a GeoSPARQL, but it's even more complicated to implement than SPARQL...).

lu-zero commented on planning/Discovery/work-items.md:
```diff
@@ -0,0 +1,51 @@
+12. Dependency - [TD Versioning](https://w3c.github.io/wot-charter-drafts/wot-wg-2023-details.html#td-versioning-workitem)
+    * Impact on Discovery includes how or whether to maintain multiple versions of a single TD, and possibly filtering by version.
+13. Dependency - [Interoperability Profiles](https://w3c.github.io/wot-charter-drafts/wot-wg-2023-details.html#profiles-workitem)
+    * Should TDDs also satisfy a Profile?
```
I think it should, Ideally one would like to subscribe to a directory and get updates either as events or as live queries (pending observable actions).

lu-zero reviewed (commented):
Here my initial feedback.

relu91 reviewed (approved):
Have minor comments. I think the work item list looks good.

mmccool commented:
Feedback from planning mtg:
- could use github projects instead 
     - can use tools to convert to/from documents (note stragegy does this)
- could use tools to generate presentations from md files
- links to liaison dependencies needed as well

JKRhb commented on planning/Discovery/work-items.md:
```diff
@@ -0,0 +1,51 @@
+14. Process - [WoT CG Coordination](https://w3c.github.io/wot-charter-drafts/wot-wg-2023-details.html#wotcg-coordination)
+    * Gathering information on requirements, implementations, feedback on current versions, suggestions for improvement, etc.
+
+## Other/New/Proposed
```
One thing I was wondering about is whether there should be some mechanism or framework for Binding Templates to define their own ways of discovering TDs.

E.g., at the moment we have explicit normative statements for HTTP and CoAP in the Discovery specification, but some of them could also be moved to their respective binding documents. Also, if a new binding like OPC-UA emerges, it would not have to make changes to the main discovery document but could define a specific mechanism on its own.

(Sorry if this has already been discussed or is already covered by the existing work items, if so then we can simply mark this as resolved.)

mmccool commented on planning/Discovery/work-items.md:
```diff
@@ -0,0 +1,51 @@
+8. Dependency - [Normative TD Parsing, Consuming and Validation](https://w3c.github.io/wot-charter-drafts/wot-wg-2023-details.html#td-consumption-workitem)
+    * Relevance: In order to support queries, TD content needs to be parsed by directory, and only valid TDs should be stored in TDDs.
+9. Dependency - [Linting mechanism for TDs](https://w3c.github.io/wot-charter-drafts/wot-wg-2023-details.html#td-linting-workitem)
+    * As above, but if meant for development may not be relevant to TDDs; to discuss.
```
Yes, I agree, it depends on the definition and purpose of "linting".  If it's just warnings and meant to provide development during feedback, then it is not critical to do it during directory registration.  However, validating a TD is useful before entering a TD into a directory, so that invalid TDs are not registered and cause problems.

lu-zero commented on planning/Discovery/work-items.md:
```diff
@@ -21,40 +21,52 @@ Prioritized list of work items related to Discovery.
+    * Assuming geolocation data added to TDs, this would add geolocation query support to Discovery.
+    * May be higher priority for some use cases - more analysis needed
+    * Requirements: [https://github.com/w3c/wot-architecture/blob/main/REQUIREMENTS/geolocation.md]
+3. Dependency - [Onboarding](https://w3c.github.io/wot-charter-drafts/wot-wg-2023-details.html#onboarding-workitem)
```
Do we want to provide few optional onboarding/pairing patterns? It could be in itself a task as big as the whole `Directory` effort. 

I'd move it to lower priority if it isn't felt necessary for one of the big use-case for Directory/Discovery.

mmccool replied:
I personally don't think "Onboarding" should be discussed as part of Discovery, it's a separate topic.  For now it has been moved down to a "Questionable" section.

mmccool commented:
We did a bunch of changes in the last Discovery call, got rid of the "Medium" priority, moved out Dependencies to their own section, split up work items with parts in two different deliverables (Geolocation), and so forth.

I'm not convinced we have the priorities right, in particular I'm not sure Geolocation should be a low priority - it depends on our target use cases, as it is a high priority for Smart City applications, for example.   But at least we have a smaller and more precise set of work items to prioritize now.

sebastiankb commented on planning/Discovery/README.md:
```diff
@@ -0,0 +1,9 @@
+# Web of Things (WoT) Discovery
```
```suggestion
# Web of Things (WoT) Discovery (Draft)
```
